src/components/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,108 @@
+import {
+  ChangeEvent,
+  ComponentProps,
+  FocusEvent,
+  MouseEvent,
+  ReactNode,
+  useEffect,
+  useState,
+} from "react"
+import styles from "./SearchInput.module.css"
+
+interface SearchInputProps extends ComponentProps<"input"> {
+  searchIcon?: ReactNode
+  clearIcon?: ReactNode
+  onDebounce?: (searchTerm: string) => void
+  debounceDelay?: number
+  onSearchClick?: (searchTerm: string) => void
+  value?: string
+}
+
+export function SearchInput({
+  name,
+  placeholder,
+  searchIcon,
+  clearIcon,
+  onDebounce,
+  debounceDelay = 300,
+  onSearchClick,
+  value = "",
+  ...inputProps
+}: SearchInputProps) {
+  const [searchTerm, setSearchTerm] = useState(value)
+  const [hideCancelButton, setHideCancelButton] = useState(true)
+
+  useEffect(() => {
+    const timer = setTimeout(() => onDebounce?.(searchTerm), debounceDelay)
+    return () => {
+      clearTimeout(timer)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [searchTerm, debounceDelay])
+
+  const handleReset = (e: MouseEvent<HTMLButtonElement>) => {
+    // we don't want the field to blur when a user clears the input
+    e.preventDefault()
+    setSearchTerm("")
+    setHideCancelButton(true)
```

Comment:
As a controlled component, the parent component will never know when the value has been reset. You'll need to call `inputProps.onChange?.(e)`. Also, `onDebounce` is called with wacky arguments after clearing.

https://user-images.githubusercontent.com/5196/184888600-00a63d53-72b8-4541-a555-73cb5e11a8ff.mov